versionCheckHook: Check for --version first

jpds opened the pull request:
More likely to find the version number in `--version` than `--help` so look for that first.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [X] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jpds committed:
versionCheckHook: Check for --version first

More likely to find the version number in `--version` than `--help` so
look for that first.

Zaczero reviewed (approved):
I went ahead and run research AI agents:

Gemini:

> The Dominant Pattern: The GNU style of --version (exclusively for metadata) and --help (exclusively for instruction) is the overwhelming norm.
> 
> The Modern Shift: A "Hybrid" pattern is emerging, primarily in Rust-based tooling, where version info is displayed in both locations. This accounts for roughly 15-20% of newer tools but does not displace the dedicated version flag.

Gemini 2:

> Based on the analysis of 108 common Linux tools, the findings are as follows:
> 
> Approximately 93% of the analyzed tools provide their version number when the --version flag (or an equivalent such as -V or -v) is used.
> 
> Approximately 53% of these tools also include their version number within the output of the --help flag (or display it when the command is invoked with invalid options or by default).

ChatGPT:

> Percentage statistics: Out of our sample of 24 tools, about 25% (roughly one-quarter) included their version number in the --help output. These were mainly interactive or user-facing tools like Bash, Wget, Apt, Vim, and Nano. In contrast, almost all tools (around 95% of those sampled) provided a dedicated --version flag (or equivalent -V) to display the version info. Only very few exceptions (≈5%) lacked a version flag (for example, the minimalist shell dash).

Based on that, I am convinced this is a positive change.

jpds committed:
treewide: Remove redundant `versionCheckProgramArg = "--version";` with:

```shell
git grep -l -e 'versionCheckProgramArg = "--version";' -e 'versionCheckProgramArg = \[ "--version" \];' | while read f; do
  sed -i '/versionCheckProgramArg/d' "$f"
  sed -i '/^$/N;/\n$/D' "$f"
done
```

doronbehar reviewed (approved):
This looks good to me, but TBH I'm not sconfident enough to merge such tree wide changes . I hope we can get a few more eyes on this.

philiptaron reviewed (approved):
I read the whole thing. LGTM.

doronbehar commented:
Tnx!

On 26 November 2025 18:26:12 GMT+02:00, Philip Taron ***@***.***> wrote:
>Merged #463218 into staging.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/463218#event-21200266847
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
